Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift

```diff
@@ -0,0 +1,274 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the swift-kafka-gsoc open source project
+//
+// Copyright (c) 2022 Apple Inc. and the swift-kafka-gsoc project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of swift-kafka-gsoc project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Crdkafka
+import Logging
+import NIOConcurrencyHelpers
+import NIOCore
+
+// TODO: write test for evaluating right state transitions
+// TODO: test abrupt run task shutdown -> cancellation handler
+
+/// A back-pressure aware polling system for managing the poll loop that polls `librdkafka` for new acknowledgements.
+final class KafkaBackPressurePollingSystem {
+    /// The element type for the system, representing either a successful ``KafkaAcknowledgedMessage`` or a ``KafkaAcknowledgedMessageError``.
+    typealias Element = Result<KafkaAcknowledgedMessage, KafkaAcknowledgedMessageError>
+    /// The producer type used in the system.
+    typealias Producer = NIOAsyncSequenceProducer<Element, NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark, KafkaBackPressurePollingSystem>
+
+    /// The state machine that manages the system's state transitions.
+    let stateMachine: StateMachine
+
+    // TODO: wrapped values are non-sendable
+    private let _client: NIOLockedValueBox<KafkaClient?>
+    private let _sequenceSource: NIOLockedValueBox<Producer.Source?>
```

Comment:
Rather than having state outside of the state machine just put it into the associated values of the state. This way they are protected by the lock. If an action has to do something with those just return it in the action.